pkg/image/controller/controller.go

```diff
@@ -29,8 +122,77 @@ func needsImport(stream *api.ImageStream) bool {
 	return stream.Annotations == nil || len(stream.Annotations[api.DockerImageRepositoryCheckAnnotation]) == 0
 }
 
-// retryCount is the number of times to retry on a conflict when updating an image stream
-const retryCount = 2
+func (c *ImportController) handleImport() {
+	for {
+		select {
+		case <-c.stopChan:
+			return
+
+		case staleImageStream := <-c.work:
+			glog.V(1).Infof("popped %s from the worklist", workingSetKey(staleImageStream))
+
+			c.importImageStream(staleImageStream)
+		}
+	}
+}
+
+func (c *ImportController) importImageStream(staleImageStream *api.ImageStream) {
+	// if we're already in the workingset, that means that some thread is already trying to do an import for this.
+	// This does NOT mean that we shouldn't attempt to do this work, only that we shouldn't attempt to do it now.
+	if !c.addToWorkingSet(staleImageStream) {
+		// If there isn't any other work in the queue, wait for a while so that we don't hot loop.
+		// Then requeue to the end of the channel.  That allows other work to continue without delay
+		if len(c.work) == 0 {
+			time.Sleep(100 * time.Millisecond)
+		}
+		glog.V(5).Infof("requeuing %s to the worklist", workingSetKey(staleImageStream))
+		c.work <- staleImageStream
+
+		return
+	}
+	defer c.removeFromWorkingSet(staleImageStream)
+
+	err := kclient.RetryOnConflict(kclient.DefaultBackoff, func() error {
+		liveImageStream, err := c.streams.ImageStreams(staleImageStream.Namespace).Get(staleImageStream.Name)
+		if err != nil {
+			return err
+		}
+		if !needsImport(liveImageStream) {
+			return nil
+		}
+
+		// if we're notified, do work and then start waiting again.
+		return c.Next(liveImageStream)
+	})
+
+	if err != nil {
```

Comment:
> do we really want to log NotFound errors as big unexpected errors?

Will explicitly deal with the common case of the error case, but if the image stream disappears as we're working on it, you'll still get a big error.  More specific handling requires more plumbing into almost-dead code.